File: grpc-server/src/main/java/ru/grpc/simple/server/grpcserver/service/UserService.java
package ru.grpc.simple.server.grpcserver.service;

import io.grpc.Status;
import io.grpc.stub.StreamObserver;
import org.apache.commons.lang3.StringUtils;
import ru.grpc.simple.proto.CommonResponse;
import ru.grpc.simple.proto.LoginRequest;
import ru.grpc.simple.proto.LogoutRequest;
import ru.grpc.simple.proto.UserGrpc;
import ru.grpc.simple.proto.util.DataUtil;

public class UserService extends UserGrpc.UserImplBase {
    private static final String ONLINE = "ONLINE";
    private static final String OFFLINE = "OFFLINE";

    @Override
    public void login(LoginRequest request, StreamObserver<CommonResponse> responseObserver) {
        String username = request.getUsername();
        String password = request.getPassword();

        CommonResponse.Builder commonResponseBuilder = CommonResponse.newBuilder();
        if (password.equals(String.valueOf(DataUtil.getCorrectPassword()))) {
            System.out.println(String.format("\"%s\"'s login attempt has succeed", username));

            DataUtil.getUsersMap().put(username, ONLINE);

            commonResponseBuilder
                    .setResponseCode(Status.Code.OK.value())
                    .setResponseMessage("Login SUCCESS");
        } else {
            System.out.println(String.format("\"%s\"'s login attempt has failed", username));

            DataUtil.getUsersMap().remove(username);

            commonResponseBuilder.setResponseCode(Status.Code.UNAUTHENTICATED.value())
                    .setResponseMessage("Login FAILED")
                    .build();
        }

        responseObserver.onNext(commonResponseBuilder.build());
        responseObserver.onCompleted();
    }

    @Override
    public void logout(LogoutRequest request, StreamObserver<CommonResponse> responseObserver) {
        String username = request.getUsername();

        CommonResponse.Builder commonResponseBuilder = CommonResponse.newBuilder();
        if (StringUtils.isEmpty(username)
                || StringUtils.isEmpty(DataUtil.getUsersMap().get(username))
                || DataUtil.getUsersMap().get(username).equals(OFFLINE)) {

            commonResponseBuilder
                    .setResponseCode(Status.Code.INVALID_ARGUMENT.value())
                    .setResponseMessage(String.format("Logout FAILED, request is empty or user \"%s\" has been OFFLINE yet",
                                                      username));
        } else {
            DataUtil.getUsersMap().put(username, OFFLINE);

            commonResponseBuilder
                    .setResponseCode(Status.Code.OK.value())
                    .setResponseMessage(String.format("Logout SUCCESS for user: \"%s\"", username));
        }

        responseObserver.onNext(commonResponseBuilder.build());
        responseObserver.onCompleted();
    }
}

File: grpc-server/src/main/kotlin/ru/grpc/simple/server/grpcserver/service/UserService.kt
package ru.grpc.simple.server.grpcserver.service

import io.grpc.Status
import ru.grpc.simple.proto.CommonResponse
import ru.grpc.simple.proto.LoginRequest
import ru.grpc.simple.proto.LogoutRequest
import ru.grpc.simple.proto.UserServiceGrpcKt
import ru.grpc.simple.proto.commonResponse
import ru.grpc.simple.server.grpcserver.repository.UserRepository
import ru.grpc.simple.server.grpcserver.repository.UserStatus

class UserService : UserServiceGrpcKt.UserServiceCoroutineImplBase() {

    override suspend fun login(request: LoginRequest): CommonResponse =
        if (UserRepository.checkPassword(request.password)) {
            println("${request.username}'s login attempt has succeed")
            UserRepository.setOnline(request.username)
            commonResponse {
                responseCode = Status.Code.OK.value()
                responseMessage = "Login SUCCESS"
            }
        } else {
            println("${request.username}'s login attempt has failed")
            UserRepository.remove(request.username)
            commonResponse {
                responseCode = Status.Code.UNAUTHENTICATED.value()
                responseMessage = "Login FAILED"
            }
        }

    override suspend fun logout(request: LogoutRequest): CommonResponse =
        if (UserRepository.getUserStatus(request.username) != UserStatus.ONLINE) {
            commonResponse {
                responseCode = Status.Code.INVALID_ARGUMENT.value()
                responseMessage = "Logout FAILED, request is empty or user ${request.username} has been OFFLINE yet"
            }
        } else {
            UserRepository.setOffline(request.username)
            commonResponse {
                responseCode = Status.Code.OK.value()
                responseMessage = "Logout SUCCESS for user: ${request.username}"
            }
        }
}